tl.tl

```diff
@@ -6346,17 +6353,41 @@ tl.type_check = function(ast: Node, opts: TypeCheckOptions): {Error}, {Error}, T
                elseif child.ktype.typename == "number" then
                   is_array = true
                   is_tuple = true
-                  if i == #children and node[i].key_parsed == "implicit" and child.vtype.typename == "tuple" then
-                     -- need to expand last item in an array (e.g { 1, 2, 3, f() })
-                     for j, c in ipairs(child.vtype) do
-                        array_len = array_len + 1
-                        node.type.elements = expand_type(node, node.type.elements, c)
-                        node.type[i+j-1] = c
+
+                  if node[i].key_parsed == "implicit" then
+                     if i == #children and child.vtype.typename == "tuple" then
+                        -- need to expand last item in an array (e.g { 1, 2, 3, f() })
+                        for j, c in ipairs(child.vtype) do
+                           node.type.elements = expand_type(node, node.type.elements, c)
+                           node.type[last_array_idx] = c
+                           last_array_idx = last_array_idx + 1
+                        end
+                     else
+                        node.type[last_array_idx] = child.vtype
+                        last_array_idx = last_array_idx + 1
+                        node.type.elements = expand_type(node, node.type.elements, child.vtype)
+                     end
+                  else -- explicit
+                     local n = node[i].key.constnum
+
+                     if not n or not is_positive_int(n) then
+                        is_map = true
+                        node.type.keys = expand_type(node, node.type.keys, child.ktype)
+                        node.type.values = expand_type(node, node.type.values, child.vtype)
+                     elseif n then
+                        if node.type[n] then
+                           node_error(node, "Overwriting index " .. n)
+                        end
+                        node.type[n] = child.vtype
+                        if n > largest_array_idx then
+                           largest_array_idx = n
+                        end
+                        node.type.elements = expand_type(node, node.type.elements, child.vtype)
```

Comment:
The overwriting error is probably something that could get demoted to a warning (if/when that becomes a thing)

> Would it be possible to continue to detect the [n] case as an array instead of a map?

Probably, I personally would classify it as a map, but it could be an array by default, as long as `{number:T}` maps are compatible with arrays

>  I suspect that a table like this [...] will result in "cannot determine type of table literal" because the [n] will trigger is_map and [2] will trigger is_array

Ah, yes that table currently results in that exact error. That should definitely be at least a map if not an array :P

Reply:
I've reverted the behavior and the above example now gets inferred as an array.
The problem this causes is that declarations like this
```
local t: {number, string} = { [-1] = "hi" }
```
are now valid, which is probably fine, but rubs me the wrong way

Reply:
> are now valid, which is probably fine, but rubs me the wrong way

There's definitely an issue here, because both `{ "hi" }` and `{ [-1] = "hi" }` print_type as `({string  "hi"})`, but one is accepted as a tuple and the other isn't. Since `{ [-1] = "hi" }` infers as an array of strings, this means that index 1 is of type string, and therefore should not be compatible with `{number, string}`. I'm taking a closer look!